### DISCUSSION TITLE
*** Format error in header line of InitIndsFile: 1 error detected

### DISCUSSION THREAD

**Discussion by Jeankwizera** [RangeShifter/RangeshiftR-tutorials#24] [Q&A] [answered]
Hi, I am trying to simulate range dispersal from an "InitIndsFile" using RangeshifR. From the Initialise( ) helpful, the InitIndsFile should be a tab separated list file with defined headers. I have a ".tsv" that looks as follows;

 Year.Species.X.Y.NInds.Sex.Age.Stage
1 0\t0\t115.6329\t-31.3291139\t1\t\t\t
2 0\t0\t145.2532\t-38.1645570\t1\t\t\t
3 0\t0\t152.0886\t-33.6075949\t1\t\t\t
4 0\t0\t154.3671\t-29.0506329\t1\t\t\t
5 0\t0\t115.6329\t-31.3291139\t1\t\t\t
6 0\t0\t145.2532\t-38.1645570\t1\t\t\t

I however get the following error:

*** Format error in header line of InitIndsFile: 1 error detected

How can I format my file to make it suitable as the "InitIndsFile"  parameter in the Initialise( ) function?

**Comment by annekathleenmalchow** [comment]
Hi,
the headers and all values need to be separated by white space.
Also, you might want to revise the information in your InitIndsFile, as each line represents a number of individuals of the same specifics (i.e.position, sex, age, ...) to be created in a certain year (for actual initialisation, this would usually be year 0 for all lines). Further, the X and Y are given as cell numbers, not as coordinates, so they must be positive integers.
Hope this helps,
Anne

**Comment by Jeankwizera** [comment]
Hi Anne, I appreciate your feedback. I adjusted the file such that all values are separated by white spaces. I also replaced the coordinates with cell numbers obtained using the colFromX( ) and rowFromY( ) functions. My column"Year" has 0 for all rows.

This is a snippet of the data:

Year	Species	lon	lat	NInds	Sex	Age	Stage
0	0	592	243	1	NA	NA	NA
0	0	651	257	1	NA	NA	NA
0	0	665	248	1	NA	NA	NA
0	0	669	239	1	NA	NA	NA
0	0	592	243	1	NA	NA	NA
0	0	651	257	1	NA	NA	NA
0	0	665	248	1	NA	NA	NA
0	0	669	239	1	NA	NA	NA
...

I however still get the error:

*** Format error in header line of InitIndsFile: 3 errors detected

**Comment by annekathleenmalchow** [comment, marked as answer]
Hi Jean,
please note the instructions given in the R help / user manual:
> The expected column headers depend on the model settings and must match the following order exactly: 'Year', 'Species' (must = 0), for cell-/patch-based: 'X', 'Y' / 'PatchID', 'Ninds', for sexual model: 'Sex', for stage-structured population: 'Age', 'Stage'. The sex is specified with 0 for female and 1 for male.

So, RangeShiftR expects that the column headers for the location are named  'X', 'Y' not  'lon', 'lat'. Further, you have to specify the entries for Sex, Age and Stage; leaving them at NA will give you the next error.
Hope this helps,
Anne

**Reply by Jeankwizera** [reply]
Hello Anne, I appreciate your feedback. I was able to work through the problem with your responses.

 I have encountered another challenge ith RunRS( ) output  i.e [*** Error in InitIndsFile at line 1: Species must be 0] which I do not understand. This is what I get:

Checking Control parameters 

Control Parameters checked

Run Simulation(s) with seed 10000 ...
LandscapeFile headers OK: /Users/kwizera.jvk/Desktop/KWIZCONSULT/RangeShiftR/Simulation_cakile_Australia/2. Climate suitability Simulation/results/Updates/UPDATED /Inputs/climate_suitability.asc
*** Error in InitIndsFile at line 1: Species must be 0
*** Did not read to EOF in InitIndsFile
Initial individuals file OK:/Users/kwizera.jvk/Desktop/KWIZCONSULT/RangeShiftR/Simulation_cakile_Australia/2. Climate suitability Simulation/results/Updates/UPDATED /Inputs/aus_ports3.tsv

Error in reading parameter file(s)... see RS log.

***** Elapsed time: 27 seconds

*****
***** Simulation completed 
***** Outputs folder: /Users/kwizera.jvk/Desktop/KWIZCONSULT/RangeShiftR/Simulation_cakile_Australia/2. Climate suitability Simulation/results/Updates/UPDATED /Outputs/
*****
$Errors
[1] 0


However, In my dataset, the values in the "Species" column are already "0"

str(read_tsv("Inputs/aus_ports3.tsv", quote = " "))
Rows: 4 Columns: 8                                                                                              
 0s── Column specification ──────────────────────────────────────────────────────────────────────────────────────────
Delimiter: "\t"
dbl (8): Year, Species, X, Y, Ninds, Sex, Age, Stage

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
spec_tbl_df [4 × 8] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ Year   : num [1:4] 0 0 0 0
 $ Species: num [1:4] 0 0 0 0
 $ X      : num [1:4] 592 651 665 669
 $ Y      : num [1:4] 243 257 248 239
 $ Ninds  : num [1:4] 1 1 1 1
 $ Sex    : num [1:4] 0 0 0 0
 $ Age    : num [1:4] 0 0 0 0
 $ Stage  : num [1:4] 0 0 0 0
 - attr(*, "spec")=
  .. cols(
  ..   Year = col_double(),
  ..   Species = col_double(),
  ..   X = col_double(),
  ..   Y = col_double(),
  ..   Ninds = col_double(),
  ..   Sex = col_double(),
  ..   Age = col_double(),
  ..   Stage = col_double()
  .. )
 - attr(*, "problems")=<externalptr> 

 
I therefore do not understand how to make "Species" satisfy the error: *** Error in InitIndsFile at line 1: Species must be 0
*** Did not read to EOF in InitIndsFile.

I would appreciate some guidance on this as well.

Regards,
Jean Kwizera

**Reply by rangeshiftr** [reply]
Hi Jean,
yes, from the output of reading your InitIndsFile within R it seems like the file format should be correct, but RS might be interpreting the file contents wrongly.
Actually, the 'Species' column should be fine but there should be an error given for the columns 'Age' and 'Stage', as they must be at least 1. (RS can't initialise age or stage 0, as initialisation happens at the beginning of the year, whereas reproduction occurs later (exact scheduling depends on your settings), and development and aging at the end. Therefore, at initialisation time, all individuals have already aged at least once.)
I am suspecting that RS might not detect the column separations correctly. Could you maybe paste the first two lines of your file here as read by a text editor? 
Alternatively, you can try to let R generate the file from the data frame that you have, for example if the data frame was named `init_df` : 
`
write.table(init_df, file=paste0(dirpath,'Inputs/aus_ports3.txt'), sep='\t', row.names=F, quote=F)
`
Best,
Anne

**Reply by Jeankwizera** [reply]
Hello, thankyou for this. The first 2 lines of the file appear as:

Year	Species	X	Y	Ninds	Sex	Age	Stage
0	0	592	243	1	0	1	1
0	0	651	257	1	0	1	1

I have made the assumption that both Age and Stage = 1 to account for RS not initialiseing age or stage 0. I have also used 
write.table(init_df, file=paste0(dirpath,'Inputs/aus_ports3.txt'), sep='\t', row.names=F, quote=F)

to write the data to file but the error (*** Error in InitIndsFile at line 1: Species must be 0) persists

**Reply by rangeshiftr** [reply]
Hi, thanks for checking this.
I assume you have also changed the file name in the Initialisation module accordingly? I cannot spot a mistake in the file that you are using, maybe the problem lies somewhere else in the simulation setup. Do you think you can create a minimal working example from your RS setup, that creates the error? Then I can test it here and dig a little deeper. You could use an Artificial Landscape for this or create a dummy landscape, so you don't need to send the landscape files. E.g. if you use habitat quality landscapes, the following could work:
```
nrow <- 300
ncol <- 600
m <- matrix(runif(nrow*ncol,1,100),nrow = nrow, ncol = ncol)

library(raster)
r <- raster(m, xmx=ncol, ymx=nrow)
writeRaster(r, format="ascii", filename = paste0(dirpath, "Inputs/toy_map"), datatype = 'INT1S', NAflag = -9L, overwrite = F)
```

**Reply by Jeankwizera** [reply]
Hello Ann, I appreciate the time you have taken to check my issue. I did try using an artificial landscape map but was still not able to by pass the same error that 

*** Error in InitIndsFile at line 1: Species must be 0

This is the sample script:

# create directories
dir.create("Inputs")
dir.create("Outputs")
dir.create("Output_Maps")

# libraries
library(raster)
library(RangeShiftR)

# create landscape
nrow <- 300
ncol <- 600
m <- matrix(runif(nrow*ncol,1,100),nrow = nrow, ncol = ncol)
r <- raster(m, xmx=ncol, ymx=nrow)
writeRaster(r, format="ascii", filename = paste0(dirpath, "Inputs/toy_map"), datatype = 'INT1S', NAflag = -9L, overwrite = F)


# Create InitIndFile
data.frame(Year = rep(0, 20),	Species = rep(0, 20),	 X = sample(592:669, 20, replace = T) ,
           Y = sample(239:257, 20, replace = T) ,	Ninds = rep(1, 20),	Sex	= rep(0, 20), 
           Age	= rep(0, 20), Stage = rep(0, 20)) |> 
  write.table(file=paste0('Inputs/locations.txt'), sep='\t', row.names=F, quote=F)

# SIMULATION

## Define Landscape for simulation
landscapes = ImportedLandscape(LandscapeFile = "toy_map.asc", 
                               Resolution = 1, 
                               HabPercent = TRUE, 
                               K_or_DensDep = 10000)

## Define Demography
demos <- Demography(Rmax = 6, bc = 0.00001, ReproductionType = 0)

## Define dispersal
dist = matrix(c(20000, 100000, 0.83 ), ncol = 3, byrow = T)
disps <-  Dispersal(Emigration = Emigration(EmigProb = 0.05), 
                    Transfer =  DispersalKernel(DoubleKernel = T, Distances = dist)) + 
  Settlement(MaxSteps = 1, Settle = 2)

## Initialize parameters including initial individuals file
inits <- Initialise(InitType = 2, # = initialisation from a loaded species distribution map
                    InitIndsFile = "locations.txt") # = all suitable cells within all distribution presence cells

# define simulation parameters
sims <- Simulation(Simulation = 1,
                   Years = 1000,
                   Replicates = 100,
                   OutIntRange = 1,
                   OutIntPop = 1,
                   OutIntOcc = 1)

## validate and simulate
simulate <- RSsim(seed = 10000) + landscapes + demos + disps + sims + inits
validateRSparams(simulate) ## check parameter validity
set.seed(23020)
RunRS(RSparams = simulate, dirpath = paste0(getwd(), "/"))

**NOTE: When I look at the Initial individuals list file in the text editor, it seems to be ok except for having " 2 white spaces in the header between the *Year* and *Species* columns. I am wondering whether that could be the source of the challenge?

**Reply by JetteReeg** [reply]
Hi Jean,

I have tested the minimal working example you provided and realized that you have a non-stage-structured, asexual model. 
Anne aleady cited the R help file above. It says that the columns "Sex", "Age" and "Stage" are only needed for stage-structured, sexual models. Which means in your case (non-stage-structured, asexual/only female) you only need to provide the columns "Year", "Species", "X", "Y" and "Ninds". I changed the location.txt file accordingly and the model runs without error message.

I hope this information helps you.

Best,
Jette

**Reply by Jeankwizera** [reply]
Hello Jette, This worked and the simulation reun without an error. I appreciate the assistance.
